projects/idea-hub-smart-contract-solidity/contracts/IdeaHub.sol

```diff
@@ -0,0 +1,288 @@
+    mapping(bytes32 => Idea) _ideas;
+
+    //// Constructor ////
+    constructor(address osoTokenAddress, address _foundationAddress) public {
+        osoToken = IERC20(osoTokenAddress);
+        foundationAddress = _foundationAddress;
+        Authenticated.manualConstruct();
+    }
+
+    // modifier
+    modifier nonZeroBytes(bytes32 _bytes){
+        require(_bytes != bytes32(0x0));
+        _;
+    }
+    //// internal helper functions ////
+    /**
+    * @dev get UserId for when address for given UserType is known
+    */
+    function _userId(address _address, uint8 userType) internal view returns (bytes32){
+        User storage user = users[_address];
+        if (user.role & userType == 0) {
+            return bytes32("");
+        }
+        return user.id;
+    }
+
+    /**
+    * @dev getUser address for given user Type when off-chain userId is known
+    */
+    function _userAddress(bytes32 id, uint8 userType) internal view returns (address){
+        address _address = userLookup[id];
+        if (_address == address(0x0)) {
+            return address(0x0);
+        }
+        User storage user = users[_address];
+        require(user.role & userType != 0);
+        return _address;
+    }
+
+    /**
+    * @dev register User with userType. Can be called multiple times to register same user for  different userTypes.
+    * @param _address Address of user
+    * @param userType User type bit to mask. only one User type at a time.
+    */
+    function _registerUser(bytes32 userId, address _address, uint8 userType) internal {
+        require(users[_address].role & userType == 0);
+        User storage user = users[_address];
+        require(userLookup[userId] == address(0x0) || userLookup[userId] == _address);
+        user.role = user.role | userType;
+        user.id = userId;
+        userLookup[userId] = _address;
+    }
+
+    /**
+    * Register an idea to platform. Can be called to update Content Hash of the idea
+    * @param ideaId hashed Id of idea
+    * @param contentHash InfoHash f
+    * @param author Address of author of idea
+    */
+    function _register(bytes32 ideaId, bytes32 contentHash, address author) internal {
+        Idea storage idea = _ideas[ideaId];
+        if (idea.author == address(0x00)) {
+            ideaIds.push(ideaId);
+            idea.author = author;
+        } else {
+            require(idea.author == author);
+        }
+        IdeaContent storage content = idea.revisions[contentHash];
+        require(!content.exists);
+        content.exists = true;
+        idea.revisionList.push(contentHash);
+        emit Registered(ideaIds.length, idea.revisionList.length);
+    }
+
+    /**
+    * @dev validate an idea
+    * @param ideaNo Sequential no obtained on registering idea
+    * @param revisionNo the revisionNo of idea to validate
+    * @param validator address of the validator
+    */
+    function _validate(uint256 ideaNo, uint256 revisionNo, address validator) internal {
```

Comment:
please add comments to make the validation procedure more clear

Reply:
Validation procedure is explained in `Readme`. Ideas are simply marked as validated in the Smart Contract and doesn't deal with the validation process itself.